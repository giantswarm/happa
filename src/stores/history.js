'use strict';

import { createBrowserHistory } from 'history';

const history = createBrowserHistory();

export default history;
