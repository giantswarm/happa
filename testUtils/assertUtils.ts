import { MatcherFunction, Nullish } from '@testing-library/react';

type Query = (matcher: MatcherFunction) => HTMLElement;

/**
 * Decorate an existing query to match text across multiple
 * elements.
 * @param query
 */
export function withMarkup(query: Query) {
  return (text: string) => {
    return query((_, node: Nullish<Element>) => {
      if (!node) return false;

      const hasText = (n: Element) => n.textContent === text;
      const hasTextInChildren = Array.from(node.children).some((child) =>
        hasText(child)
      );

      return hasText(node) && !hasTextInChildren;
    });
  };
}
