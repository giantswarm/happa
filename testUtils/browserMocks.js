// eslint-disable-next-line @typescript-eslint/no-var-requires
const fetch = require('isomorphic-fetch');

const localStorageMock = (function () {
  let store = {};

  return {
    getItem: function (key) {
      return store[key] || null;
    },
    setItem: function (key, value) {
      store[key] = value;
    },
    removeItem: function (key) {
      delete store[key];
    },
    clear: function () {
      store = {};
    },

    // replaceWith is not part of the localStorage spec, but useful for us
    // in testing to set the full localStorage to a known set of values.
    replaceWith: function (newStore) {
      store = newStore;
    },
  };
})();

Object.defineProperty(window, 'localStorage', {
  value: localStorageMock,
});

// Adjust the console logging behaviour during test runs.
/* eslint-disable no-console */
global.console = {
  log: console.log,
  error: console.error,
  warn: jest.fn(), // warnings are surpressed.
  info: console.info,
  debug: console.debug,
};
/* eslint-enable no-console */

global.fetch = fetch;
