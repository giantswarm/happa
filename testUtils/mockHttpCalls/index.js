/***** Constants *****/
export * from './constantsAndHelpers';

/***** Responses *****/

// Info
export * from './info';

// User
export * from './user';

// Auth
export * from './auth';

// Clusters
export * from './clusters';

// Organizations
export * from './organizations';

// Status
export * from './status';

// Apps
export * from './apps';

// Node Pools
export * from './nodePools';

// Releases
export * from './releases';

// App Catalogs
export * from './appCatalogs';
