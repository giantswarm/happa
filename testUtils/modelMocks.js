import * as authSelectors from 'selectors/authSelectors';

jest.mock('model/services/giantSwarm');
jest.mock('selectors/authSelectors');

// eslint-disable-next-line no-import-assign
authSelectors.selectAuthToken = jest.fn(
  (_, state) => state.app.loggedInUser.token
);
