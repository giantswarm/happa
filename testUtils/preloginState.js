import { Providers } from 'shared/constants';
import { MainRoutes } from 'shared/constants/routes';

// This is what the state looks like when someone brand new arrives at the site
// and they are not logged in at all.

export default {
  router: {
    location: {
      pathname: MainRoutes.Login,
      search: '',
      hash: '',
      key: '0fgz24',
    },
    action: 'POP',
  },
  main: {
    selectedOrganization: null,
    selectedClusterID: null,
    firstLoadComplete: false,
    loggedInUser: null,
    permissions: {},
    info: {
      general: {
        installation_name: '',
        availability_zones: null,
        provider: Providers.AWS,
      },
      stats: {
        cluster_creation_duration: {
          median: 0,
          p25: 0,
          p75: 0,
        },
      },
      workers: {
        count_per_cluster: {
          max: null,
          default: 0,
        },
      },
    },
  },
  errors: {},
  metadata: {
    version: {
      current: '',
      new: null,
      lastCheck: 0,
      timer: 0,
    },
  },
  entities: {
    catalogs: {
      lastUpdated: 0,
      isFetching: false,
      items: {},
    },
    clusters: {
      lastUpdated: 0,
      isFetching: false,
      items: {},
      v5Clusters: [],
      idsAwaitingUpgrade: {},
    },
    organizations: {
      lastUpdated: 0,
      isFetching: false,
      items: {},
      credentials: {
        lastUpdated: 0,
        isFetching: false,
        items: [],
        showForm: false,
      },
    },
    releases: {
      error: null,
      isFetching: false,
      items: {},
    },
    users: {
      lastUpdated: 0,
      isFetching: false,
      items: {},
      invitations: {
        lastUpdated: 0,
        isFetching: false,
        items: {},
      },
    },
    nodePools: {
      items: {},
      isFetching: false,
    },
    clusterLabels: {
      requestInProgress: false,
      error: null,
    },
  },
  modal: {
    visible: false,
    templateValues: {},
    template: '',
  },
};
