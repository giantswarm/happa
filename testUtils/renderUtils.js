import { render } from '@testing-library/react';
import App from 'App';
import { ConnectedRouter } from 'connected-react-router';
import { ThemeProvider } from 'emotion-theming';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Provider } from 'react-redux';
import { AppRoutes } from 'shared/constants/routes';
import configureStore from 'stores/configureStore';
import theme from 'styles/theme';

export const initialStorage = {
  user:
    '{"email":"developer@giantswarm.io","auth":{"scheme":"giantswarm","token":"a-valid-token"},"isAdmin":true}',
};

/**
 * This function will render the whole app with a mocked store in the route
 * provided.
 * @param {String} route.
 * @param {HTMLElement} container An HTMLElement.
 */
export function renderRouteWithStore(
  initialRoute = AppRoutes.Home,
  state = {},
  storage = initialStorage
) {
  localStorage.replaceWith(storage);

  const history = createMemoryHistory({
    initialEntries: [initialRoute],
    initialIndex: 0,
  });

  const store = configureStore(state, history);

  const app = render(<App {...{ store, theme, history }} />);

  return app;
}

/**
 * Render component in the Theme context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithTheme(component, props, options) {
  return render(getComponentWithTheme(component, props), options);
}

/**
 * Include component in the Theme context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 */
export function getComponentWithTheme(Component, props) {
  return (
    <ThemeProvider theme={theme}>
      <Component {...props} />
    </ThemeProvider>
  );
}

/**
 * Render component in the Store context
 * @param {React.ReactType} component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {<Q extends TestingLibrary.Queries>TestingLibrary.RenderOptions<Q>} [options] Testing library render options
 */
export function renderWithStore(component, props, state, options) {
  return render(getComponentWithStore(component, props, state), options);
}

/**
 * Include component in the Store context and return it
 * @param {React.ReactType} Component The React Component
 * @param {<P extends Record<string, any>>React.ComponentProps<P>} props Props to pass to the component
 * @param {Record<string, any>} state Current Store state
 * @param {Record<string, any>} storage Current LocalStorage value
 * @param {History} history Current Browser history
 */
export function getComponentWithStore(
  Component,
  props = {},
  state = {},
  storage = initialStorage,
  history = createMemoryHistory()
) {
  localStorage.replaceWith(storage);

  const store = configureStore(state, history);

  const app = (
    <Provider store={store}>
      <ThemeProvider theme={theme}>
        <ConnectedRouter history={history}>
          <Component {...props} />
        </ConnectedRouter>
      </ThemeProvider>
    </Provider>
  );

  return app;
}
