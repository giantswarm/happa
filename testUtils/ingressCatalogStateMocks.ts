import { Constants } from 'shared';
import { appCatalogsResponse } from 'testUtils/mockHttpCalls';

const ingressCatalog = appCatalogsResponse.find(
  (c) => c.metadata.name === Constants.INSTALL_INGRESS_TAB_APP_CATALOG_NAME
);
export const catalogsState = {
  entities: {
    catalogs: {
      items: {
        [Constants.INSTALL_INGRESS_TAB_APP_CATALOG_NAME]: ingressCatalog,
      },
    },
  },
};
