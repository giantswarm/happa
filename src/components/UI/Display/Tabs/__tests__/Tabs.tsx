import { fireEvent, render, screen } from '@testing-library/react';
import { createMemoryHistory } from 'history';
import React from 'react';
import { getComponentWithStore, initialStorage } from 'testUtils/renderUtils';

import { Tab, Tabs } from '..';

describe('Tabs', () => {
  const renderComponent = (
    props: Partial<React.ComponentProps<typeof Tabs>> = {
      children: [],
    },
    history = createMemoryHistory()
  ) => {
    return render(
      getComponentWithStore(Tabs, props, {}, initialStorage, history)
    );
  };

  it('renders without crashing', () => {
    renderComponent();
  });

  it('renders only the first tab', () => {
    renderComponent({
      children: [
        <Tab key='tab-1' title='first'>
          <h1>First Tab</h1>
        </Tab>,
        <Tab key='tab-2' title='second'>
          <h1>Second Tab</h1>
        </Tab>,
        <Tab key='tab-3' title='third'>
          <h1>Third Tab</h1>
        </Tab>,
      ],
    });

    expect(screen.queryByText(/first tab/i)).toBeInTheDocument();
    expect(screen.queryByText(/second tab/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/third tab/i)).not.toBeInTheDocument();
  });

  it(`changes the URL when switching tabs, if the 'useRoutes' option is enabled`, () => {
    const history = createMemoryHistory();

    renderComponent(
      {
        children: [
          <Tab key='tab-1' path='/' title='First tab'>
            <h1>First Content</h1>
          </Tab>,
          <Tab key='tab-2' path='/second-tab' title='Second Tab'>
            <h1>Second Content</h1>
          </Tab>,
          <Tab key='tab-3' path='/third-tab' title='Third Tab'>
            <h1>Third Content</h1>
          </Tab>,
        ],
        useRoutes: true,
      },
      history
    );

    fireEvent.click(screen.getByText(/second tab/i));
    expect(history.location.pathname).toBe('/second-tab');

    fireEvent.click(screen.getByText(/third tab/i));
    expect(history.location.pathname).toBe('/third-tab');

    fireEvent.click(screen.getByText(/first tab/i));
    expect(history.location.pathname).toBe('/');
  });

  it('renders a basic tab layout', () => {
    const { container } = renderComponent({
      children: [
        <Tab key='tab-1' title='first'>
          <h1>First Tab</h1>
        </Tab>,
        <Tab key='tab-2' title='second'>
          <h1>Second Tab</h1>
        </Tab>,
        <Tab key='tab-3' title='third'>
          <h1>Third Tab</h1>
        </Tab>,
      ],
    });

    expect(container.firstChild).toMatchSnapshot();
  });
});
