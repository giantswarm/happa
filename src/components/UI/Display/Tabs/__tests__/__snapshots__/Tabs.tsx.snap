// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs renders a basic tab layout 1`] = `
<div
  class="StyledGrommet-sc-19lkkz7-0 kQNwCm"
>
  <div
    class="ThemeProvider__AppWrapper-sc-12xd2cf-0 fmpnGZ"
  >
    <div
      class="StyledBox-sc-13pk1d4-0 jsevgO StyledTabs-sc-a4fwxl-2 dqMpcz"
    >
      <div
        class="StyledBox-sc-13pk1d4-0 dHsKBb"
        role="tablist"
      >
        <div
          class="StyledBox-sc-13pk1d4-0 eRCQcV StyledTabs__StyledTabsHeader-sc-a4fwxl-0 doyXrw"
        >
          <button
            aria-expanded="true"
            aria-selected="true"
            class="StyledButtonKind-sc-1vhfpnt-0 gUiHYe"
            role="tab"
            type="button"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 kAiAch StyledTab-sc-1nnwnsb-0 cwUFig"
            >
              <span
                class="StyledText-sc-1sadyjn-0 dPgXjd"
              >
                first
              </span>
            </div>
          </button>
          <button
            aria-expanded="false"
            aria-selected="false"
            class="StyledButtonKind-sc-1vhfpnt-0 gUiHYe"
            role="tab"
            type="button"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 ehaPYh StyledTab-sc-1nnwnsb-0 cwUFig"
            >
              <span
                class="StyledText-sc-1sadyjn-0 ksyYGh"
              >
                second
              </span>
            </div>
          </button>
          <button
            aria-expanded="false"
            aria-selected="false"
            class="StyledButtonKind-sc-1vhfpnt-0 gUiHYe"
            role="tab"
            type="button"
          >
            <div
              class="StyledBox-sc-13pk1d4-0 ehaPYh StyledTab-sc-1nnwnsb-0 cwUFig"
            >
              <span
                class="StyledText-sc-1sadyjn-0 ksyYGh"
              >
                third
              </span>
            </div>
          </button>
        </div>
      </div>
      <div
        aria-label="first Tab Contents"
        class="StyledTabs__StyledTabPanel-sc-a4fwxl-1 edURMz"
        role="tabpanel"
      >
        <h1>
          First Tab
        </h1>
      </div>
    </div>
  </div>
</div>
`;
